File: examples/chatroom/bin/post.js
// chatroom verb: post a message to the room this verb shipped with.
//
// Invoked from a shell that mounted the room: `room:post hello world`.
// Runs CONFINED (ADR 0007 §Confinement contract): the namespace is exactly
// the room at /res plus stdio and argv — this code cannot see anything else.
//
// Input convention: argv joined with spaces is the message body; with no argv
// the body is read from stdin, so `echo hi | room:post` composes.
//
// Self-contained by design: a /bin verb resolves imports against the confined
// namespace root, so it must not import local libraries.
import * as std from "qjs:std";

const args = scriptArgs.slice(1);
const body = args.length ? args.join(" ") : std.in.readAsString();
if (!body) {
  std.err.puts("post: empty message (pass words or pipe a body)\n");
  std.exit(1);
}
const file = std.open("/res/post", "w");
if (!file) {
  std.err.puts("post: cannot open /res/post\n");
  std.exit(1);
}
file.puts(body);
file.close();

File: crates/wanix-wasm/fixtures/verbs-src/src/post.rs
//! The `post` verb: a resource-shipped command that posts a message to the
//! resource it came from.
//!
//! Input convention (documented in docs/site/content/concepts/bin-verbs.md):
//! the argv words joined with spaces are the message body; with no argv the
//! body is read from stdin (so `echo hi | room:post` composes in a pipeline).
//! The verb runs confined: its namespace is exactly the resource at `/res`,
//! so the post lands on `/res/post` and nowhere else can be touched.

use std::io::{Read, Write};

fn main() {
    let args: Vec<String> = std::env::args().skip(1).collect();
    let body = if args.is_empty() {
        let mut bytes = Vec::new();
        if let Err(err) = std::io::stdin().read_to_end(&mut bytes) {
            eprintln!("post: stdin: {err}");
            std::process::exit(1);
        }
        bytes
    } else {
        args.join(" ").into_bytes()
    };
    // The guest file exists by declaration: open write-only without
    // create/truncate (O_CREAT|O_TRUNC is refused on device-like files).
    let result = std::fs::OpenOptions::new()
        .write(true)
        .open("/res/post")
        .and_then(|mut file| file.write_all(&body));
    if let Err(err) = result {
        eprintln!("post: /res/post: {err}");
        std::process::exit(1);
    }
}
